caddyhttp/staticfiles/fileserver.go

```diff
@@ -80,17 +82,34 @@ func (fs FileServer) serveFile(w http.ResponseWriter, r *http.Request, name stri
 	}
 
 	// redirect to canonical path
-	url := r.URL.Path
 	if d.IsDir() {
-		// Ensure / at end of directory url
-		if !strings.HasSuffix(url, "/") {
-			Redirect(w, r, path.Base(url)+"/", http.StatusMovedPermanently)
+		// Ensure / at end of directory url. If the original URL path is
+		// used then ensure / exists as well.
+		if !strings.HasSuffix(r.URL.Path, "/") {
+			toURL, _ := url.Parse(r.URL.String())
+
+			path, ok := r.Context().Value(caddy.URLPathContextKey).(string)
```

Comment:
Hmm, do we want to use the context's path value here?

A Caddyfile like:

```
example.com/subpath
root /www/mysite
```

And then a request to `example.com/subpath/test.html` should load `/www/mysite/test.html` -- is that what you would expect? I think that's how it's currently implemented, anyway.